<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2025-4</h1>

<h2> Profesor: Sergio A. Alcalá Corona </h2> 

---
### Ayudante: Rodrigo Vega Vilchis
---

---
### Alumno: Germán Olivares Domínguez
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 1.  Programación básica


### Programación básica para la física

---

### **EJERCICIOS**:

$\;$

#### **1. Altitud de un satélite**: se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos.

#### **(a)** Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es:

<h3 align="center">$h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R$,</h3> 

donde $G = 6,67 × 10^{−11} m^3 kg^{−1}s^{−2}$ es la constante gravitatoria de Newton, $M = 5,97 ×
1024 kg$ es la masa de la Tierra, y $R = 6371 km$ es su radio.




<div style="background-color:#f0f8ff; padding:10px; border-radius:5px;">
El satélite se mueve en un movimiento circular uniforme. La única fuerza que actúa sobre él es la gravitacional, que también funciona como fuerza centrípeta y curva su trayectoria. Por lo tanto, la fuerza gravitacional y la fuerza centrípeta son iguales.

Escribimos la ley de gravitación de Newton y la fórmula de la fuerza centrípeta:

Ley de gravitación de Newton:

$$
F_g = \frac{G m M}{(R + h)^2},
$$

donde \( m \) representa la masa del satélite y \( h \) es la altitud del satélite sobre la superficie de la Tierra.

La fuerza centrípeta es igual a:

$$
F_c = m a_d = \frac{m v^2}{R + h},
$$

donde \( v \) es la velocidad tngencial del satélite.

Ambas fuerzas deben ser iguales, por lo tanto:

$$
\frac{G m M}{(R + h)^2} = \frac{m v^2}{R + h} \tag{1}
$$

El satélite es geoestacionario, por lo que su velocidad angular es igual a la velocidad angular de la Tierra. Así, el periodo de este movimiento es también el mismo: 24 horas.

Entonces, la velocidad tangencial del satélite es:

$$
v = \omega (R + h) = \frac{2 \pi}{T} (R + h), \tag{2}
$$

donde \( T \) representa su periodo.

Usando las fórmulas (1) y (2) podemos expresar la altura:

La velocidad angular en la fórmula (2) se sustituye en la fórmula (1):

$$
\frac{G m M}{(R + h)^2} = \frac{4 \pi^2 m (R + h)}{T^2} \implies \frac{G M T^2}{4 \pi^2} = (R + h)^3
$$

La solución final para \( h \) es:

$$
h = \sqrt[3]{\frac{G M T^2}{4 \pi^2}} - R
$$
</div>


#### **(b)** Escribe un programa que le pida al usuario que ingrese el valor deseado de $T$ y luego calcule e imprima la altura correcta en metros.


In [3]:
import math
G = 6.67e-11   
M = 5.97e24  
R = 6371e3
T = float(input("Ingresa el periodo orbital T en segundos: "))
h = (G * M * T**2 / (4 * math.pi**2))**(1/3) - R
print("La altitud del satélite es:", h, "metros")

Ingresa el periodo orbital T en segundos: 17893509
La altitud del satélite es: 1471751700.6297257 metros


#### **(c)** Utiliza tu programa para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada *órbita "geoestacionaria"*), una vez cada **90 minutos** y una vez cada **45 minutos**. 
#### ¿Qué concluyes del último de estos cálculos?

In [4]:
print("\nAltitudes para: ")
T_dia = 24 * 3600
T_90min = 90 * 60
T_45min = 45 * 60

lista_T = [T_dia, T_90min, T_45min]
nombres = ["Geoestacionaria (24h)", "Cada 90 minutos", "Cada 45 minutos"]

for i in range(3):
    T = lista_T[i]
    h = (G * M * T**2 / (4 * math.pi**2))**(1/3) - R
    print(nombres[i], "Altitud:", h, "metros")


Altitudes para: 
Geoestacionaria (24h) Altitud: 35855910.176174976 metros
Cada 90 minutos Altitud: 279321.6253728606 metros
Cada 45 minutos Altitud: -2181559.8978108233 metros


Se concluye que para un periodo orbital demasiado corto, el satelite no puede completar una orbita circular. Es decir, a periodos pequeños, el satelite se encuentra "dentro de la Tierra", ya que R es mayor a r, por eso la altitud da negativa.

#### **2. Relatividad especial:** Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 

#### Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

#### **(a)** en el marco de reposo de un observador en la Tierra y

In [5]:
import math
c = 1.0  # c: velocidad de la luz dada en unidades de años luz por año
x = float(input("Ingresa la distancia al planeta (en años luz): "))
c_fraccion = float(input("Ingresa la velocidad como fracción decimal de c: "))
v = c_fraccion * c
if v >= c:
    print("La velocidad debe ser menor que la velocidad de la luz.")
else:
    t_tierra = x / v
    t_nave = t_tierra * math.sqrt(1 - v**2 / c**2)
    print("Tiempo desde un observador en la Tierra:", t_tierra, "años")

Ingresa la distancia al planeta (en años luz): 23
Ingresa la velocidad como fracción decimal de c: 0.6
Tiempo desde un observador en la Tierra: 38.333333333333336 años


#### **(b)** como lo percibiría un pasajero a bordo de la nave. 

In [6]:
import math
c = 1.0
x = float(input("Ingresa la distancia al planeta (en años luz): "))
c_fraccion = float(input("Ingresa la velocidad como fracción decimal de c: "))
v = c_fraccion * c
if v >= c:
    print("La velocidad debe ser menor que la velocidad de la luz.")
else:
    t_tierra = x / v
    t_nave = t_tierra * math.sqrt(1 - v**2 / c**2)
    print("Tiempo según un pasajero en la nave:", t_nave, "años")

Ingresa la distancia al planeta (en años luz): 35
Ingresa la velocidad como fracción decimal de c: 0.9
Tiempo según un pasajero en la nave: 16.951273669324838 años


#### Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

In [7]:
x = 10
v = 0.99
t_tierra = x / v
t_nave = t_tierra * math.sqrt(1 - v**2)
print("Tiempo en Tierra:", t_tierra, "años")
print("Tiempo en la nave:", t_nave, "años")

Tiempo en Tierra: 10.1010101010101 años
Tiempo en la nave: 1.424922826228878 años


#### **3. Órbitas planetarias**: En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su \textit{perihelio}), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

#### **(a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o \textit{afelio}, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3>
       
#### donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

#### Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.


<div style="background-color:#f0f8ff; padding:10px; border-radius:5px;">
De la segunda ley de Kepler se cumple que:

$$
\ell_1 v_1 = \ell_2 v_2
$$

Considerando la conservación de la energía total:

$$
\frac{1}{2} v_1^2 - \frac{G M}{\ell_1} = \frac{1}{2} v_2^2 - \frac{G M}{\ell_2}
$$

Al sustituir 

$$
\ell_2 = \frac{\ell_1 v_1}{v_2}
$$

queda:

$$
\frac{1}{2} v_1^2 - \frac{G M}{\ell_1} = \frac{1}{2} v_2^2 - \frac{G M v_2}{\ell_1 v_1}
$$

Reordenando, se obtiene la ecuación cuadrática:

$$
v_2^2 - \frac{2 G M}{\ell_1 v_1} v_2 - \left( v_1^2 - \frac{2 G M}{\ell_1} \right) = 0
$$

De aquí, las soluciones para \(v2\) son:

$$
v_2 = \frac{ \frac{2 G M}{\ell_1 v_1} \pm \sqrt{ \left( \frac{2 G M}{\ell_1 v_1} \right)^2 + 4 \left( v_1^2 - \frac{2 G M}{\ell_1} \right) } }{2}
$$

Se escoge la raíz menor porque la velocidad en el afelio debe ser menor que en el perihelio.


</div>


#### **(b)**  Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center"><b>Semieje mayor</b>: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2),$</h3>

<h4 align="center"><b>Semieje menor</b>: $\; \; b = \sqrt{\ell_1\ell_2},$</h3>

<h4 align="center"><b>Período orbital</b>: $\; \; T = {2\pi ab\over\ell_1 v_1},$</h3>

<h4 align="center"><b>Excentricidad orbital</b>: $\; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1},$</h3>

#### Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades $\ell_2$, $v_2$, $T$ y $e$.


In [10]:
import math
G = 6.6738e-11
M = 1.9891e30
l1 = float(input("Ingresa la distancia al Sol en el perihelio (en metros): "))
v1 = float(input("Ingresa la velocidad en el perihelio (en m/s): "))

# Calculamos el discriminante para obtener v2 usando conservacion de la energía y conservación del momento angular
# Conservación de energía:
# 0.5*v1^2 - GM/l1 = 0.5*v2^2 - GM/l2
# Y usando l1*v1 = l2*v2 entonces v2 = l1*v1 / l2
# Se obtiene una cuadrática en v2
A = 1
B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)
discriminante = B**2 - 4*A*C
if discriminante < 0:
    print("No hay solución real para v2.")
else:
    raiz = math.sqrt(discriminante)
    v2_1 = (-B + raiz) / (2*A)
    v2_2 = (-B - raiz) / (2*A)
    
    # Tomamos la raíz menor positiva
    if v2_1 < v2_2 and v2_1 > 0:
        v2 = v2_1
    else:
        v2 = v2_2

    # Calcular l2
    l2 = l1 * v1 / v2

    # Semiejes
    a = 0.5 * (l1 + l2)
    b = math.sqrt(l1 * l2)

    # Periodo orbital
    T = (2 * math.pi * a * b) / (l1 * v1)

    # Excentricidad
    e = (l2 - l1) / (l2 + l1)
    
    print("l2 =", l2, "m")
    print("v2 =", v2, "m/s")
    print("Periodo orbital T =", T, "s")
    print("Excentricidad e =", e)

Ingresa la distancia al Sol en el perihelio (en metros): 123465789
Ingresa la velocidad en el perihelio (en m/s): 12346
l2 = 123465789.0 m
v2 = 12346.0 m/s
Periodo orbital T = 62834.799237334766 s
Excentricidad e = 0.0


#### **(c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [11]:
print("Valores para la Tierra:")
l1 = 1.4710e11
v1 = 3.0287e4

B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)
discriminante = B**2 - 4*A*C
v2 = (-B - math.sqrt(discriminante)) / (2*A)
l2 = l1 * v1 / v2
a = 0.5 * (l1 + l2)
b = math.sqrt(l1 * l2)
T = (2 * math.pi * a * b) / (l1 * v1)
e = (l2 - l1) / (l2 + l1)
print("T (Tierra) =", T / (3600*24*365), "años")

print("Valores para el cometa Halley:")
l1 = 8.7830e10
v1 = 5.4529e4

B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)
discriminante = B**2 - 4*A*C
v2 = (-B - math.sqrt(discriminante)) / (2*A)
l2 = l1 * v1 / v2
a = 0.5 * (l1 + l2)
b = math.sqrt(l1 * l2)
T = (2 * math.pi * a * b) / (l1 * v1)
e = (l2 - l1) / (l2 + l1)
print("T (Halley) =", T / (3600*24*365), "años")

Valores para la Tierra:
T (Tierra) = 1.0002238777234564 años
Valores para el cometa Halley:
T (Halley) = 76.08170065465461 años


### 4. La fórmula semiempírica de la masa (FSM)

En física nuclear, la **fórmula de Weizsäcker** (conocida también como **fórmula semiempírica**) sirve para evaluar la masa y otras propiedades de un núcleo atómico, y está basada parcialmente en mediciones empíricas.
En particular, la fórmula se usa para calcular la **energía de enlace nuclear aproximada** \( B \), de un núcleo atómico con número atómico \( Z \) y número de masa \( A \):

$$
B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A - 2Z)^2}{A} + \frac{a_5}{\sqrt{A}}
$$

donde, en unidades de millones de electron-volts (MeV), las constantes son:

$$
a_1 = 15.8, \quad a_2 = 18.3, \quad a_3 = 0.714, \quad a_4 = 23.2
$$

y


$$
a_5 =
\begin{cases}
0 & \text{si } A \text{ es impar} \\
12.0 & \text{si } A \text{ y } Z \text{ son pares (ambos)} \\
-12.0 & \text{si } A \text{ es par y } Z \text{ impar}
\end{cases}
$$
---



a) Escribe un programa que tome como entrada los valores de A y Z, e imprima la 
energía de enlace B para el átomo correspondiente. Usa tu programa para encontrar
la energía de enlace de un átomo con A = 58 y Z = 28 (Hint: La respuesta
correcta es alrededor de los 490 MeV).


In [12]:
import math as mt
import numpy as np
import matplotlib.pyplot as pl

A = int(input("Ingresa el número de masa A: "))
Z = int(input("Ingresa el número atómico Z: "))

def EnergiaEnlace(A,Z):
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    a5 = 0

    if (A % 2 == 0) and (Z % 2 == 0):
        a5 = 12
    elif (A % 2 == 0) and (Z % 2 != 0):
        a5 = -12 
    elif (A % 2 != 0):
        a5 = 0

    B = a1*A - a2*(A**(2/3)) - a3*((Z**2)/(A**(1/3))) - a4*(((A-2*Z)**2)/A)+ a5/(A**(1/2))
    return B
B = EnergiaEnlace(A, Z)
print("Energía de enlace B =", B, "MeV")

Ingresa el número de masa A: 58
Ingresa el número atómico Z: 28
Energía de enlace B = 497.5620206224374 MeV


b) Modifica el programa del inciso anterior, para escribir una segunda versión que
imprima no la energía de enlace total B, sino la energía de unión por nucleón, que
es B/A.


In [13]:
import math as mt
import numpy as np
import matplotlib.pyplot as pl

A = int(input("Ingresa el número de masa A: "))
Z = int(input("Ingresa el número atómico Z: "))

def EnergiaEnlace(A,Z):
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    a5 = 0

    if (A % 2 == 0) and (Z % 2 == 0):
        a5 = 12
    elif (A % 2 == 0) and (Z % 2 != 0):
        a5 = -12 
    elif (A % 2 != 0):
        a5 = 0

    B = a1*A - a2*(A**(2/3)) - a3*((Z**2)/(A**(1/3))) - a4*(((A-2*Z)**2)/A)+ a5/(A**(1/2))
    return B
B = EnergiaEnlace(A, Z)        
print("Energía de unión por nucleón (B/A) =", B/A, "MeV")

Ingresa el número de masa A: 58
Ingresa el número atómico Z: 28
Energía de unión por nucleón (B/A) = 8.578655527973059 MeV


c) Escribe una tercera versión del programa para que tome como entrada solo un
valor del número atómico \( Z \) y luego pase por todos los valores de \( A \) desde \( A = Z \)
hasta \( A = 3Z \), para encontrar el que tiene la mayor energía de enlace por nucleón.
Este es el núcleo más estable con el número atómico dado. Haz que tu programa
imprima el valor de \( A \) para este núcleo más estable y el valor de la energía de
enlace por nucleón.

---


In [14]:
import math as mt
import numpy as np
import matplotlib.pyplot as pl

Z = int(input("Ingresa el número atómico Z para buscar A más estable: "))

def EnergiaEnlace3(Z):
    Energias = []
    As = []

    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2

    for A in range(Z, 3*Z + 1):    
        if (A % 2 == 0) and (Z % 2 == 0):
            a5 = 12
        elif (A % 2 == 0) and (Z % 2 != 0):
            a5 = -12 
        else:
            a5 = 0

        B = a1*A - a2*(A**(2/3)) - a3*((Z**2)/(A**(1/3))) - a4*(((A - 2*Z)**2)/A) + a5/(A**(1/2))
        B_por_A = B / A

        Energias.append(B_por_A)
        As.append(A)

    # Encontrar índice del valor máximo
    i_max = np.argmax(Energias)
    A_mas_estable = As[i_max]
    Bmax = Energias[i_max]

    print("El núcleo más estable con Z =", Z, "es A =", A_mas_estable)
    print("Energía de enlace por nucleón máxima:", Bmax, "MeV")

# Llamar a la función
EnergiaEnlace3(Z)


Ingresa el número atómico Z para buscar A más estable: 28
El núcleo más estable con Z = 28 es A = 62
Energía de enlace por nucleón máxima: 8.70245768367189 MeV


d) Finalmente, escribe una cuarta versión del programa que, en lugar de tomar \( Z \)
como entrada, se ejecute a través de todos los valores de \( Z \) de 1 a 100 e imprima
el valor más estable de \( A \) para cada uno. ¿A qué valor de \( Z \) se produce la energía
de enlace máxima por nucleón? (La respuesta correcta, en la vida real, es \( Z = 28 \),
que corresponde al Níquel).

In [15]:
import math as mt
import numpy as np

def EnergiaEnlace4():

    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    a5 = 0

    Emax = []      # guardará la energía máxima por nucleón para cada Z
    Aestables = [] # guardará el A más estable para cada Z

    for Z in range(1, 101):

        Energias = []
        Avals = []

        for A in range(Z, 3*Z + 1):

            if (A % 2 == 0) and (Z % 2 == 0):
                a5 = 12
            elif (A % 2 == 0) and (Z % 2 != 0):
                a5 = -12
            else:  # si A es impar
                a5 = 0

            B = a1 * A - a2 * (A ** (2/3)) - a3 * ((Z ** 2) / (A ** (1/3))) - a4 * (((A - 2*Z) ** 2) / A) + a5 / (A ** (1/2))
            energia_por_nucleon = B / A
            Energias.append(energia_por_nucleon)
            Avals.append(A)

        maximo = max(Energias)
        indice_max = Energias.index(maximo)
        A_mas_estable = Avals[indice_max]

        Emax.append(maximo)
        Aestables.append(A_mas_estable)

        print("Z =", Z, "A más estable =", A_mas_estable, "Energía máxima por nucleón =", round(maximo, 4), "MeV")

    energia_global_max = max(Emax)
    indice_global_max = Emax.index(energia_global_max)
    Z_energia_max = indice_global_max + 1
    A_energia_max = Aestables[indice_global_max]

    print("\n-----")
    print("Valor de Z con energía de enlace máxima por nucleón: Z =", Z_energia_max)
    print("A más estable para ese Z: A =", A_energia_max)
    print("Energía máxima por nucleón:", round(energia_global_max, 4), "MeV")

    return energia_global_max, Z_energia_max, A_energia_max


# Ejecutar función
EnergiaEnlace4()


Z = 1 A más estable = 3 Energía máxima por nucleón = 0.3687 MeV
Z = 2 A más estable = 4 Energía máxima por nucleón = 5.3219 MeV
Z = 3 A más estable = 7 Energía máxima por nucleón = 5.2802 MeV
Z = 4 A más estable = 8 Energía máxima por nucleón = 6.4663 MeV
Z = 5 A más estable = 11 Energía máxima por nucleón = 6.6501 MeV
Z = 6 A más estable = 14 Energía máxima por nucleón = 7.2009 MeV
Z = 7 A más estable = 15 Energía máxima por nucleón = 7.3309 MeV
Z = 8 A más estable = 18 Energía máxima por nucleón = 7.7193 MeV
Z = 9 A más estable = 19 Energía máxima por nucleón = 7.737 MeV
Z = 10 A más estable = 22 Energía máxima por nucleón = 8.0354 MeV
Z = 11 A más estable = 25 Energía máxima por nucleón = 8.0256 MeV
Z = 12 A más estable = 26 Energía máxima por nucleón = 8.2412 MeV
Z = 13 A más estable = 29 Energía máxima por nucleón = 8.241 MeV
Z = 14 A más estable = 30 Energía máxima por nucleón = 8.3792 MeV
Z = 15 A más estable = 33 Energía máxima por nucleón = 8.3852 MeV
Z = 16 A más estable = 36

(8.70245768367189, 28, 62)

### 5. La constante de Madelung

En física de la materia condensada, la **constante de Madelung** da el potencial eléctrico
total que siente un átomo en un sólido; y depende de las cargas de los otros átomos
cercanos y de sus ubicaciones.

Por ejemplo, el cristal de cloruro de sodio sólido (la sal de mesa), tiene átomos dispuestos
en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio
una carga positiva $+e$ y los de cloro una carga negativa $-e$ (donde $e$ es la carga del electrón).

Si etiquetamos cada posición en la red con tres coordenadas enteras $(i, j, k)$, entonces
los átomos de sodio caen en posiciones donde $i + j + k$ es par, y los átomos de cloro en
posiciones donde $i + j + k$ es impar.

Consideremos un átomo de sodio en el origen, i.e. $i = j = k = 0$, y calculemos la
constante de Madelung. Si el espaciado de los átomos en la red es $a$, entonces la distancia
desde el origen al átomo en la posición $(i, j, k)$ es:

$$
\sqrt{(ia)^2 + (ja)^2 + (ka)^2} = a \sqrt{i^2 + j^2 + k^2},
$$

y el potencial en el origen creado por tal átomo es:

$$
V(i, j, k) = \pm \frac{e}{4 \pi \varepsilon_0 a \sqrt{i^2 + j^2 + k^2}},
$$

siendo $\varepsilon_0$ la permitividad del vacío y el signo de la expresión se toma dependiendo de
si $i + j + k$ es par o impar.

Así entonces, el potencial total que siente el átomo de sodio
es la suma de esta cantidad sobre todos los demás átomos. Supongamos una caja cúbica
alrededor del átomo de sodio en el origen, con $L$ átomos en todas las direcciones, entonces:

$$
V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ (i,j,k)\ne(0,0,0)}}^{L} V(i, j, k) = \frac{e}{4 \pi \varepsilon_0 a} M,
$$

donde $M$ es la constante de Madelung (al menos aproximadamente).

Técnicamente, la constante de Madelung es el valor de $M$ cuando $L \to \infty$, pero se
puede obtener una buena aproximación simplemente usando un valor grande de $L$.

**Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro
de sodio. Utiliza un valor de $L$ tan grande como puedas, sin dejar que tu programa se
ejecute por más de un minuto.**

In [16]:
import math

L = int(input("Ingresa el valor de L (tamaño de la red cúbica): "))
M = 0.0 # Inicializar la constante de Madelung

# Triple ciclo para recorrer todos los puntos en la caja cúbica
for i in range(-L, L + 1):
    for j in range(-L, L + 1):
        for k in range(-L, L + 1):
            # Saltar el origen (0, 0, 0)
            if i == 0 and j == 0 and k == 0:
                continue

            r = math.sqrt(i**2 + j**2 + k**2)
            signo = (-1) ** (i + j + k)
            M += signo / r

print(f"Constante de Madelung aproximada (L):", L," M = ", M)


Ingresa el valor de L (tamaño de la red cúbica): 20
Constante de Madelung aproximada (L): 20  M =  -1.719401169453936


### 6. Coeficientes binomiales

El coeficiente binomial $\binom{n}{k}$ es un número entero igual a:

$$
\binom{n}{k} = \frac{n!}{k!(n - k)!} = \frac{n \times (n - 1) \times (n - 2) \times \dots \times (n - k + 1)}{1 \times 2 \times \dots \times k}
$$

donde $k \geq 1$, o bien $\binom{n}{0} = 1$ cuando $k = 0$.

**a)** Utiliza esta fórmula para escribir una función llamada `binomial(n, k)` (o como tú quieras) que calcule el coeficiente binomial para un $n$ y $k$ dados. Asegúrate de
que tu función devuelva la respuesta en forma de un número entero (no flotante)
y proporcione el valor correcto de $1$ para el caso en que $k = 0$.

In [15]:
import math

n = int(input("Ingresa n: "))
k = int(input("Ingresa k: "))

num = 1
for i in range(n, n - k, -1):
    num *= i

den = 1
for i in range(1, k + 1):
    den *= i

binomio = num // den
print("El coeficiente binomial (", n, "sobre", k, ") es:", binomio)

Ingresa n: 2
Ingresa k: 5
El coeficiente binomial ( 2 sobre 5 ) es: 0


**b)** Usando tu función, escribe un programa que imprima las primeras 20 líneas del
“triángulo de Pascal”. La $n$-ésima línea del triángulo de Pascal contiene $n+1$ números,
que son los coeficientes $\binom{n}{0}$, $\binom{n}{1}$, y así sucesivamente hasta $\binom{n}{n}$. 

De tal manera que las primeras líneas son:

<pre>
               
             1   1
           1   2   1
         1   3   3   1
       1   4   6   4   1
    
</pre>


In [4]:
import math
print("\nTriángulo de Pascal:")
for n in range(1,21):
    fila = ""
    for k in range(n + 1):
        num = 1
        for i in range(n, n - k, -1):
            num *= i
        den = 1
        for i in range(1, k + 1):
            den *= i
        fila += str(num // den) + " "
    print(fila)



Triángulo de Pascal:
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 
1 10 45 120 210 252 210 120 45 10 1 
1 11 55 165 330 462 462 330 165 55 11 1 
1 12 66 220 495 792 924 792 495 220 66 12 1 
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1 
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1 
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1 
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1 
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1 
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1 
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1 
1 20 190 1140 4845 15504 38760 77520 125970 167960 184756 167960 125970 77520 38760 15504 4845 1140 190 20 1 


**c)** La probabilidad de que, para una moneda no sesgada lanzada $n$ veces, salga águila
$k$ veces es:

$$
p(k \mid n) = \frac{\binom{n}{k}}{2^n}
$$

Escribe un programa para calcular:

1) La probabilidad total de que una moneda lanzada 100 veces, salga águila exactamente 60 veces.

In [17]:
import math
n = 100
k = 60

# Calcular binomio
num = 1
for i in range(n, n - k, -1):
    num *= i

den = 1
for i in range(1, k + 1):
    den *= i

binom = num // den
p = binom / 2**n
print("Probabilidad de obtener exactamente 60 águilas:", p)


Probabilidad de obtener exactamente 60 águilas: 0.010843866711637987


2) La probabilidad total de que salga águila **60 veces o más**.

In [58]:
import math
prob = 0.0
for k in range(60, 101):
    num = 1
    for i in range(n, n - k, -1):
        num *= i
    den = 1
    for i in range(1, k + 1):
        den *= i
    binom = num // den
    prob += binom / 2**n

print("Probabilidad de obtener 60 o más águilas:", prob)


Probabilidad de obtener 60 o más águilas: 0.028443966820490392


### 7. Recursión

Una característica muy útil de las funciones que podemos definir es la **recursividad**, es
decir, la capacidad de una función de llamarse a sí misma. 

Por ejemplo, considere la siguiente definición del factorial $n!$ de un entero positivo $n$:

$$
n! =
\begin{cases}
1 & \text{si } n = 1, \\
n \times (n - 1)! & \text{si } n > 1.
\end{cases}
$$

Esto constituye una definición completa del factorial que nos permite calcular el valor
de $n!$ para cualquier entero positivo. Podemos emplear esta definición directamente
para crear una función para calcular factoriales, como esta (en Python):

```
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)
```
Nota cómo si $n$ no es igual a 1, la función se llama a sí misma para calcular el factorial
de $n - 1$, de tal manera que al pedir "print(factorial(5))" el intérprete de Python imprimirá correctamente la respuesta 120.

**a)** Anteriormente, en clase encontramos los números de Catalán $C_n$. Si hacemos una
pequeña modificación, la definición de $C_n$ se puede reescribir de la siguiente forma:

$$
C_n = 
\begin{cases}
1 & \text{if } n = 0, \\
\dfrac{4n - 2}{n + 1} C_{n-1} & \text{if } n > 0.
\end{cases}
$$

Escribe una función, usando recursividad, que calcule $C_n$ y utilízala para calcular
e imprimir $C_{100}$.

In [60]:
import math
def catalan(n):
    if n == 0:
        return 1
    else:
        return ((4*n - 2) * catalan(n - 1)) // (n + 1)

print("C_100 =", catalan(100))


C_100 = 896519947090131496687170070074100632420837521538745909320


**b)** Euclides demostró que el máximo común divisor $g(m, n)$ de dos enteros no negativos
$m$ y $n$ satisface que:

$$
g(m, n) =
\begin{cases}
m & \text{si } n = 0, \\
g(n, m \bmod n) & \text{si } n > 0.
\end{cases}
$$

Usando esta fórmula, escribe una función $g(m,n)$ que emplee recursividad para
calcular el máximo común divisor de $m$ y $n$. Usa tu función para calcular e imprimir
el máximo común divisor de 108 y 192.

In [62]:
import math
m = 108
n = 192

while n != 0:
    r = m % n
    m = n
    n = r

print("Máximo común divisor (MCD) de 108 y 192 =", m)


Máximo común divisor (MCD) de 108 y 192 = 12


Comparando el cálculo de los números de Catalán del inciso 7a, con lo hecho en clase,
vemos que es posible hacer el cálculo de dos maneras, ya sea directamente o usando
recursividad.

En la mayoría de los casos, si una cantidad se puede calcular sin recursividad, será
más rápido y normalmente se recomienda hacerlo así, si es posible. 

Sin embargo, hay algunos cálculos que son esencialmente imposibles (o al menos mucho más difíciles)
sin recursividad.